FIX #17: stored notebook's directory to aid client in finding stored file

jrmims opened the pull request:
I believe I have fixed issue 17 with the understanding that the user will not call os.chdir() before creating a SigPlot object. Doing so breaks the path that is stored as "notebook_dir" on creation.

I've ran through a few different scenarios and not had issues, but please double check my work with existing notebooks.

I removed the TODO [here](https://github.com/LGSInnovations/jupyter-sigplot/compare/master...jrmims:issue_17?expand=1#diff-cde14e61eb0a0fead146831ff9271ba1L398) because I believe I have taken care of it.

sterre commented on jupyter_sigplot/sigplot.py:
```diff
@@ -61,12 +61,11 @@ def __init__(self, *args, **kwargs):
+        self.data_dir = kwargs.pop('data_dir', self.notebook_dir)
+        if not os.path.isabs(self.data_dir):
+            # Need to store absolute path of data directory in case user os.chdir
+            self.data_dir = os.getcwd() + "/" + self.data_dir
```
Suggest using `unravel_path` here, or at least `os.path.join`, instead of explicit string manipulation

sterre commented:
I added one comment against the code in the PR, if we want to go in that direction.

On further reflection, if we can say that `self.data_dir` is always an absolute path when specified in the constructor, then I don't think we need the other changes that pass `notebook_dir` down to internal helper functions. We could even guarantee this by using a property for `data_dir` so that we always store it as an absolute path, regardless of how the user specifies it.

But once you start thinking "we only store absolute paths for data_dir," another perspective arises:  we could just document `data_dir` as always being relative to the notebook server's cwd (or the kernel's cwd, whichever one is true). You can imagine cases where this might be desired, and if it's documented, user surprise would be minimal. 

As I recall, my original observation was because I noticed that the notebook server's cwd was not necessarily the parent directory of a given notebook. There was something in the code at that time that resulted in an error. It may have been addressed by #18. The important distinction may be *kernel cwd* vs. *notebook server cwd*, rather than `chdir` as I was thinking when I wrote the comment. 

As an aside, one of my personal use cases is to set `data_dir` to something like `cache` or `datafiles` so that I can easily specify any cached files in `.gitignore`, without ignoring files that I might include explicitly with my notebook. This has its own wrinkles, independent of `chdir`. 

So anyway, I am starting to think that just pointing out the subtlety of kernel vs. notebook server where `data_dir` is defined/documented may give the user the greatest flexibility. It also removes concern about timing from the extension itself.

Let me know what you think!

jrmims commented:
> I added one comment against the code in the PR, if we want to go in that direction.
> 
> On further reflection, if we can say that `self.data_dir` is always an absolute path when specified in the constructor, then I don't think we need the other changes that pass `notebook_dir` down to internal helper functions. We could even guarantee this by using a property for `data_dir` so that we always store it as an absolute path, regardless of how the user specifies it.
> 
> But once you start thinking "we only store absolute paths for data_dir," another perspective arises: we could just document `data_dir` as always being relative to the notebook server's cwd (or the kernel's cwd, whichever one is true). You can imagine cases where this might be desired, and if it's documented, user surprise would be minimal.
> 
> As I recall, my original observation was because I noticed that the notebook server's cwd was not necessarily the parent directory of a given notebook. There was something in the code at that time that resulted in an error. It may have been addressed by #18. The important distinction may be _kernel cwd_ vs. _notebook server cwd_, rather than `chdir` as I was thinking when I wrote the comment.
> 
> As an aside, one of my personal use cases is to set `data_dir` to something like `cache` or `datafiles` so that I can easily specify any cached files in `.gitignore`, without ignoring files that I might include explicitly with my notebook. This has its own wrinkles, independent of `chdir`.
> 
> So anyway, I am starting to think that just pointing out the subtlety of kernel vs. notebook server where `data_dir` is defined/documented may give the user the greatest flexibility. It also removes concern about timing from the extension itself.
> 
> Let me know what you think!

the passing of `notebook_dir` was due to the possibility of a user changing directories after a SigPlot object was created. In this situation, the kernel's directory changes, and so the relative path from kernel cwd to the data is broken.
```
import os
plot = SigPlot("data/sin.tmp")
os.chdir("..")
plot.plot()
```
if we can state in the documentation that users are not supposed to os.chdir(), then we can store the `data_dir` as a relative path. I think in that situation, the current master branch works without this PR.

sterre commented:
> if we can state in the documentation that users are not supposed to os.chdir(), then we can store the `data_dir` as a relative path. I think in that situation, the current master branch works without this PR.

Exactly. Users can create their own use cases if we don't get in their way too much. And it's not even "don't cd"--more like, "be aware that changing the kernel's working directory will affect the interpretation of `data_dir` if `data_dir` is specified as a relative path".

I think the original problem that triggered my comment/issue was that the extension itself was inconsistent, which we have since addressed. 

Here's an example of where cd might make sense for an analytic session:

```python
In[1]
def showdata():
  plot = SigPlot("data/sin.tmp")
  plot.plot()
  return plot
```

```python
In[2]
os.chdir("/path/to/orig_files")
showdata()
```

```python
In[3]
os.chdir("/path/to/filtered_files")
showdata()
```

So, because I believe the original problem is OBE, and we will serve our users better by informing them than by trying to predict their use cases, I believe we can just update the docs for `data_dir` as above and call this one good.

jrmims commented:
I believe I have captured everything discussed in this PR. I rolled back the first series of commits, fixed the flake8 test, and added documentation in sigplot.py and the example notebook. Let me know if there are any more changes you'd like to see!

sterre commented:
I like those comments, @jrmims . As far as I'm concerned, the PR as it stands addresses the issue.
